app/controllers/IdentityController.scala

```diff
@@ -0,0 +1,33 @@
+package controllers
+
+import actions.CommonActions._
+import com.gu.i18n.CountryGroup
+import controllers.forms.MetadataUpdate
+import models._
+import monitoring._
+import play.api.mvc._
+import services.IdentityService
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class IdentityController(identityService: IdentityService)(implicit ec: ExecutionContext)
+  extends Controller with Redirect with TagAwareLogger with LoggingTagsProvider {
+  val metadataUpdateForm = MetadataUpdate.metadataUpdateForm
+
+  def updateMarketingPreferences(countryGroup: CountryGroup) = NoCacheAction.async(parse.form(metadataUpdateForm)) { implicit request =>
+    val marketingOptIn = request.body.marketingOptIn
+    val contributor = request.session.data.get("email") match {
+      case Some(email) => identityService.updateMarketingOptIn(email, marketingOptIn)
+      case None => Future.successful(error("email not found in session while trying to update marketing opt in"))
+    }
+
+    val url = request.session.get("amount").flatMap(ContributionAmount.apply)
+      .filter(_ => request.isAndroid)
+      .map(mobileRedirectUrl)
+      .getOrElse(routes.Contributions.thanks(countryGroup).url)
+
+    contributor.map { _ =>
```

Comment:
I don't think you need to use `map` here, or even create `contributor`. Just evaluating the `match` statement will be enough.